### ISSUE TITLE
Automate Test Coverage

### ISSUE THREAD

**Issue by davethenipper** [synergistic-silobusters/flintandsteel#225]
Acceptance Criteria:
- [ ] To Come

We should not be delivering code that does not have test coverage.  It would be great if our tools could detect if coverage drops and fail the build.

**Comment by Mctalian** [comment]
Bear with me, here, a lot of this is just steam of consciousness...

So I thought about this one a bit. We should be able to generate coverage results in an additional format, like .txt or something else that can be easily parsed. We would have a script that runs at the end of the build to parse the coverage data for comparison. Then we need to set up a small database, possibly a Redis database since it focuses on key value pairs- it is super simple. This database needs to be accessible by our CI builds. Basically each build on master, we will overwrite some coverage data in that database. Each PR will compare coverage numbers to the most recent build on master and make sure it's at least equal, preferably greater, and it will return success. If the coverage numbers fall, it will fail (possibly just flag the build somehow? We'll need to look into it).

Alternatively, we can set a static threshold for coverage. Possibly 80%. The script could just check if we've fallen below 80% and fail then. That way it's less strict, in case we have problems covering some tests and we fall by like .1% and the build fails.

I'm actually feeling option 2 now, doesn't need a database, it just keeps us honest with our PRs.

**Comment by LongArmMcGee** [comment]
Option 2 is more appealing. We are a small team and we can easily
self-regulate. If we can quickly get a small script working then that's
great! But we don't need to invest much time into an effort like this
because of our LEAN AGILEity.

We can look at adding a script to bower to check coverage as well. We could
have it throw a non-critical error if coverage drops, so that the user is
made aware, but uninhibited. This would cut effort associated with our CI
systems. I assume adding things to our CI builds is more time consuming?

TL;DR: An easy solution would be great, let's not drop much time into this.

**Comment by YashdalfTheGray** [comment]
Adding something to the CI build makes sure that the developer is aware of it because the build fails. I'm always up for self-regulation because that makes you learn more but I think adding it to the CI build is definitely a requirement. 

Travis even has Code Climate code coverage options that we can set and if we use Istanbul, which I'm almost certain that we do (the reports look like it), we can have Travis automatically parse coverage results and report them!

**Comment by Mctalian** [comment]
Wasn't aware of code climate. Without doing any searching myself that
sounds like it has potential!

**Comment by YashdalfTheGray** [comment]
Well I'm familiar with Code Climate but from a ruby side from Nationwide. I don't know if it works inherently with JavaScript but travis' version of Code Climate can parse Istanbul code coverage output and know what's going on.

**Comment by Mctalian** [comment]
One thing I noticed on my fork's master is that code-climate marks my build failed if any new issues arise and it is extremely picky. It only seems to run coverage on master and not on PRs as well.
